### PR TITLE
Offline mode implementation

### PR DESCRIPTION
**Pull request by Tsathogguaa**
Making necessary preparation to implement the offline usage of the app. Mostly done. Some of the code that is added is in the form of commentaries, as the required parts to compile those lines have yet to be implemented.

Have to discuss test conditions with the team.

### CODE REVIEW

**Comment by Tsathogguaa** [comment]
> Making necessary preparation to implement the offline usage of the app. Mostly done. Some of the code that is added is in the form of commentaries, as the required parts to compile those lines have yet to be implemented.
> 
> Have to discuss test conditions with the team.

Quick note about why the code in DownloadManager is commented and not tested: 

The tests would require both a download and an API call, both of them are not stable things to test. If the LyricsBelongGameActivity tests can be injected somehow, then I can suggest testing this piece of code as well, otherwise the tests may fail very often, both because of the API and the download activity. Since our coverage is low and I currently do not need this piece of code(not until next week's PR), I left the code commented to not decrease the coverage for no use.

File: app/src/main/java/ch/sdp/vibester/helper/BuzzerGameManager.kt [hunk 1]
```diff
@@ -22,7 +23,15 @@ class BuzzerGameManager: GameManager() {
...
-        mediaPlayer = AudioPlayer.playAudio(currentSong.getPreviewUrl())
+        if(hasInternet) {
+            mediaPlayer = AudioPlayer.playAudio(currentSong.getPreviewUrl())
+        } else {
+            mediaPlayer = AudioPlayer.playAudio(currentSong.getPreviewUrl())
+            /* TODO: OFFLINE
+            var media = File(context.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS), "extract_of_$currentSong.getTrackName()")
+            mediaPlayer = AudioPlayer.playAudio(media.absolutePath)
+             */
+        }
```
Comment (zwierski) [inline app/src/main/java/ch/sdp/vibester/helper/BuzzerGameManager.kt:34]: Could you just put these modifications into GameManager instead? I deleted BuzzerGameManager when refactoring and put all the equivalent functions into GameManager instead so you should find it there

Comment (Tsathogguaa) [reply app/src/main/java/ch/sdp/vibester/helper/BuzzerGameManager.kt:34]: Done on local, will push soon once I make sure the merge was successful

File: app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            if (method == LastfmMethod.BY_ARTIST.method) {
+                tracksField = "toptracks"
+            }
+
+            fillList()
+
+            page = "1"
+            songsPerPage = "100"
+            totalPages = "20"
+            totalSongs = "2000"
```
[the remark this answers is not in the document]
Comment (Tsathogguaa) [reply app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt:37]: True, will do it

File: app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    /**
+     * Adds from the downloads to the list of songs ("$songName", "$artistName")
+     * Saves the list of songs in songList
+     */
+    private fun fillList() {
+        var records = File(context.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS), "records.txt")
+
+
+        if(!records.exists() || records.length() == 0L) {
```
Comment (zwierski) [inline app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt:52]: Add a space for readability `if (!records.exists() || records.length() == 0L)`

Comment (Tsathogguaa) [reply app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt:52]: Done

File: app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            //There are no downloaded songs, keep the song list empty
+            emptySongs = true
+        } else {
+            var reader = BufferedReader(FileReader(records))
+            var currentLine = reader.readLine()
+
+            while(currentLine != null) {
+                var trimmed = currentLine.trim()
+                if(trimmed.isNotEmpty()) {
```
Comment (zwierski) [inline app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt:61]: Add a space after if

File: app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                    val split = trimmed.split("-")
+                    if(split.size == 2) {
```
Comment (zwierski) [inline app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt:63]: Add a space after if

File: app/src/main/java/ch/sdp/vibester/activity/DownloadActivity.kt [hunk 6]
```diff
@@ -204,8 +204,30 @@ class DownloadActivity : AppCompatActivity() {
...
 
+    /**
+     * Records the properties of a song.
+     * Order of storage: Track name - artist name - artwork URL - preview URL.
+     */
+    /* TODO: OFFLINE
+    private fun recordProperties() {
+        var properties = File(applicationContext.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS), "properties.txt")
+
+        if(!properties.exists()) {
```
Comment (zwierski) [inline app/src/main/java/ch/sdp/vibester/activity/DownloadActivity.kt:220]: Add a space after if

**Review by zwierski — COMMENTED** [review]
Good job! Commented on a few quick readability changes.

**Comment by Tsathogguaa** [comment]
> Good job! Commented on a few quick readability changes.

Went over them and committed the changes. Can you quickly check if they are better now?

**Review by zwierski — APPROVED** [review]
LGTM 👍

File: app/src/androidTest/java/ch/sdp/vibester/model/OfflineSongListTest.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        val context = InstrumentationRegistry.getInstrumentation().targetContext.applicationContext
+        val mySongsList = OfflineSongList(context)
+
+        val inputSongsList = mutableListOf<Pair<String, String>>()
+        val page = "1"
+        val songsPerPage = "100"
+        val totalPages = "20"
+        val totalSongs = "2000"
+
+
```
Comment (laurislopata) [inline app/src/androidTest/java/ch/sdp/vibester/model/OfflineSongListTest.kt:23]: extra space

File: app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    private val page = "1"
+    private val songsPerPage = "100"
+    private val totalPages = "20"
+    private val totalSongs = "2000"
+    private var context = ctx
+    private var emptySongs: Boolean = false
+
+    init {
+        try {
+
```
Comment (laurislopata) [inline app/src/main/java/ch/sdp/vibester/model/OfflineSongList.kt:27]: extra spaces

**Review by laurislopata — APPROVED** [review]
LGTM, just some minor readability comments

**Pull request merged**